### PR TITLE
Fix missing scroll back when stop in Song Editor

### LINKED ISSUE #2145: Song editor scroll back
A little feature missing in the song editor:
The last button "After stopping go back to begin" does what is supposed to do (the marker is back) but the song editor does not scroll back so you have to scroll manually to the beginning of the song.
I think this would be useful.

### PR DESCRIPTION
**Pull request by midi-pascal**
fixes #2145 via meta-issue 2291
Make the Song Editor scroll back to the beginning of song or to the time line position depending on what is selected with the 3 states stop button.

### CODE REVIEW

**Comment by Umcaruje** [comment]
@midi-pascal I tested this one out. It now scrolls back to the beginning to the song editor, but regardless if autoscroll is on or off. I think that the song editor shouldn't scroll back when autoscroll is off.

**Comment by midi-pascal** [comment]
@Umcaruje Oups! I will check that.

**Comment by midi-pascal** [comment]
@Umcaruje I tested this and I think my patch works correctly since the stetting that allow to scroll back - or not - is designed to act when stopping the play back.
What I see is:
- If the auto scroll is off
 - If the scroll back setting is "After stopping keep position" there is no scroll back (I think this is what you meant here)
 - If the setting is one of the other two the song editor scrolls back to where it should (beginning of the song or to the time line respectively). I think it is what it is supposed to do IMHO.
- If the auto scroll is on, the setting works as expected when stopping the playback

So I think the way it works with my patch is OK unless I misunderstood something in the way the scroll back should work.
Let me know what you think about this.

**Comment by tresf** [comment]
@Umcaruje do you agree?  If so we can merge.  If not, we should fix.  Scrollback behavior has been less than ideal for a very long time, this would be a nice fix to see for 1.2. 👍

**Comment by midi-pascal** [comment]
:+1:

**Comment by musikBear** [comment]
@midi-pascal :+1:

**Comment by midi-pascal** [comment]
@Umcaruje Any advice on my explanation? This PR would be nice for 1.2 IMHO :wink:

**Comment by Umcaruje** [comment]
Sorry, I overlooked the mail. 

My opinion is that when autoscroll is off, the song editor should never scroll, in any case. What autoscroll really means is 'follow the playhead'. And as a user, when I turn it off, I expect that the song editor never scrolls back to where the playhead is, regardless of its behavior.

>If the setting is one of the other two the song editor scrolls back to where it should (beginning of the song or to the time line respectively).

The thing is, those buttons alter how the *playhead* behaves, and not the song editor. It should be independent.

**Comment by midi-pascal** [comment]
@Umcaruje I think I found the proper fix to avoid unwanted Song Editor scrolling. As usual, your feedback is very welcome :smile:

**Comment by Umcaruje** [comment]
@midi-pascal this now works as expected :+1: 

Though, here's a thing that's inconsistent (this is nitpicking though): With autoscroll on, when I press stop, it does not scroll back smoothly, as it does when e.g. I move the timeline, and then press play with the playhead at the beginning. You have to enable smooth scrolling from the performance menu in the settings for this.

**Comment by midi-pascal** [comment]
@Umcaruje So... I think this new fix should do the trick for Smooth Scroll. I tried all the the test cases I could think of and it seems OK to me. (no scroll, "hard" scroll, smooth scroll, and all the combinations with the "were to scroll or not").
I must say I never used the smooth scroll before and this is a nice feature I must say :+1: ).
You are of a very precious help to me. Thanks!
Waiting for your advice, now.

File: src/gui/editors/SongEditor.cpp [hunk 1]
```diff
@@ -568,11 +568,7 @@ void SongEditor::updatePosition( const MidiTime & _t )
 		}
 		else if( _t < m_currentPosition )
 		{
-			MidiTime t = qMax(
-				(int)( _t - w * MidiTime::ticksPerTact() /
-							pixelsPerTact() ),
-									0 );
-			animateScroll( m_leftRightScroll, t.getTact(), m_smoothScroll );
+			animateScroll( m_leftRightScroll, _t.getTact(), m_smoothScroll );
```
Comment (Umcaruje) [inline src/gui/editors/SongEditor.cpp]: This underscored function parameter should be renamed to just `t`, as that is no longer a coding convention.

**Comment by Umcaruje** [comment]
Ok, tested this out, works perfectly :+1: Great work.

You just need to adress that one styling remark, and this'll be good to merge.

**Comment by midi-pascal** [comment]
@Umcaruje Being at it, I removed the leading underscores throughout the SongEditor fonctions parameters so it is done once for all. :wink:

File: src/gui/editors/SongEditor.cpp [hunk 2]
```diff
@@ -313,54 +313,54 @@ void SongEditor::setEditModeSelect()
...
-		QWidget::keyPressEvent( _ke );
+		QWidget::keyPressEvent( ke );
 	}
 }
 
 
 
 
-void SongEditor::wheelEvent( QWheelEvent * _we )
+void SongEditor::wheelEvent(QWheelEvent * we )
```
Comment (Umcaruje) [inline src/gui/editors/SongEditor.cpp]: Why remove the space here?

File: src/gui/editors/SongEditor.cpp [hunk 3]
```diff
@@ -382,22 +382,22 @@ void SongEditor::wheelEvent( QWheelEvent * _we )
...
 		return;
 	}
-	_we->accept();
+	we->accept();
 }
 
 
 
-void SongEditor::closeEvent( QCloseEvent * _ce )
+void SongEditor::closeEvent(QCloseEvent * ce )
```
Comment (Umcaruje) [inline src/gui/editors/SongEditor.cpp]: Why remove the space here?

File: src/gui/editors/SongEditor.cpp [hunk 4]
```diff
@@ -457,9 +457,9 @@ void SongEditor::hideMasterVolumeFloat( void )
 
 
 
-void SongEditor::setMasterPitch( int _new_val )
+void SongEditor::setMasterPitch(int new_val )
```
Comment (Umcaruje) [inline src/gui/editors/SongEditor.cpp]: And here? (there are a couple of other places, but you get the point).

**Comment by Umcaruje** [comment]
>@Umcaruje Being at it, I removed the leading underscores throughout the SongEditor fonctions parameters so it is done once for all. :wink:

:+1:

**Comment by midi-pascal** [comment]
@Umcaruje Sorry, my Qt Creator settings did that while refactoring. Spaces are back now. :smiley:

**Comment by midi-pascal** [comment]
@Umcaruje So?

**Comment by Umcaruje** [comment]
Ok, did a last test, all looks good. Merging.

**Comment by Umcaruje** [comment]
Thanks for your work on this @midi-pascal :+1:

**Pull request merged**